File: sources/Android/android-communications/library/src/main/java/com/polar/androidcommunications/api/ble/model/gatt/client/pmd/model/AccData.kt
package com.polar.androidcommunications.api.ble.model.gatt.client.pmd.model

import com.polar.androidcommunications.api.ble.model.gatt.client.pmd.BlePMDClient
import com.polar.androidcommunications.api.ble.model.gatt.client.pmd.BlePMDClient.Companion.parseDeltaFramesAll
import com.polar.androidcommunications.api.ble.model.gatt.client.pmd.PmdDataFrame
import com.polar.androidcommunications.common.ble.TypeUtils

internal class AccData {
    data class AccSample internal constructor(
        val timeStamp: ULong,
        // Sample contains signed x,y,z axis values in milliG
        val x: Int,
        val y: Int,
        val z: Int
    )

    val accSamples: MutableList<AccSample> = mutableListOf()

    companion object {
        private const val TYPE_0_SAMPLE_SIZE_IN_BYTES = 1
        private const val TYPE_0_SAMPLE_SIZE_IN_BITS = TYPE_0_SAMPLE_SIZE_IN_BYTES * 8
        private const val TYPE_0_CHANNELS_IN_SAMPLE = 3

        private const val TYPE_1_SAMPLE_SIZE_IN_BYTES = 2
        private const val TYPE_1_SAMPLE_SIZE_IN_BITS = TYPE_1_SAMPLE_SIZE_IN_BYTES * 8
        private const val TYPE_1_CHANNELS_IN_SAMPLE = 3

        private const val TYPE_2_SAMPLE_SIZE_IN_BYTES = 3
        private const val TYPE_2_SAMPLE_SIZE_IN_BITS = TYPE_2_SAMPLE_SIZE_IN_BYTES * 8
        private const val TYPE_2_CHANNELS_IN_SAMPLE = 3

        fun parseDataFromDataFrame(frame: PmdDataFrame): AccData {
            return if (frame.isCompressedFrame) {
                when (frame.frameType) {
                    PmdDataFrame.PmdDataFrameType.TYPE_0 -> dataFromCompressedType0(frame)
                    PmdDataFrame.PmdDataFrameType.TYPE_1 -> dataFromCompressedType1(frame)
                    else -> throw java.lang.Exception("Compressed FrameType: ${frame.frameType} is not supported by ACC data parser")
                }
            } else {
                when (frame.frameType) {
                    PmdDataFrame.PmdDataFrameType.TYPE_0 -> dataFromRawType0(frame)
                    PmdDataFrame.PmdDataFrameType.TYPE_1 -> dataFromRawType1(frame)
                    PmdDataFrame.PmdDataFrameType.TYPE_2 -> dataFromRawType2(frame)
                    else -> throw java.lang.Exception("Raw FrameType: ${frame.frameType} is not supported by ACC data parser")
                }
            }
        }

        private fun dataFromRawType0(frame: PmdDataFrame): AccData {
            val accData = AccData()
            var offset = 0
            val step = TYPE_0_SAMPLE_SIZE_IN_BYTES
            val samplesSize = frame.dataContent.size / (step * TYPE_0_CHANNELS_IN_SAMPLE)
            val timeStamps = PmdTimeStampUtils.getTimeStamps(previousFrameTimeStamp = frame.previousTimeStamp, frameTimeStamp = frame.timeStamp, samplesSize = samplesSize, frame.sampleRate)
            var timeStampIndex = 0

            while (offset < frame.dataContent.size) {
                val x = TypeUtils.convertArrayToSignedInt(frame.dataContent, offset, step)
                offset += step
                val y = TypeUtils.convertArrayToSignedInt(frame.dataContent, offset, step)
                offset += step
                val z = TypeUtils.convertArrayToSignedInt(frame.dataContent, offset, step)
                offset += step
                accData.accSamples.add(AccSample(timeStamp = timeStamps[timeStampIndex], x = x, y = y, z = z))
                timeStampIndex++
            }
            return accData
        }

        private fun dataFromRawType1(frame: PmdDataFrame): AccData {
            val accData = AccData()
            var offset = 0
            val step = TYPE_1_SAMPLE_SIZE_IN_BYTES
            val samplesSize = frame.dataContent.size / (step * TYPE_1_CHANNELS_IN_SAMPLE)
            val timeStamps = PmdTimeStampUtils.getTimeStamps(previousFrameTimeStamp = frame.previousTimeStamp, frameTimeStamp = frame.timeStamp, samplesSize = samplesSize, frame.sampleRate)
            var timeStampIndex = 0

            while (offset < frame.dataContent.size) {
                val x = TypeUtils.convertArrayToSignedInt(frame.dataContent, offset, step)
                offset += step
                val y = TypeUtils.convertArrayToSignedInt(frame.dataContent, offset, step)
                offset += step
                val z = TypeUtils.convertArrayToSignedInt(frame.dataContent, offset, step)
                offset += step
                accData.accSamples.add(AccSample(timeStamp = timeStamps[timeStampIndex], x = x, y = y, z = z))
                timeStampIndex++
            }
            return accData
        }

        private fun dataFromRawType2(frame: PmdDataFrame): AccData {
            val accData = AccData()
            var offset = 0

            val step = TYPE_2_SAMPLE_SIZE_IN_BYTES

            val samplesSize = frame.dataContent.size / (step * TYPE_2_CHANNELS_IN_SAMPLE)
            val timeStamps = PmdTimeStampUtils.getTimeStamps(previousFrameTimeStamp = frame.previousTimeStamp, frameTimeStamp = frame.timeStamp, samplesSize = samplesSize, frame.sampleRate)
            var timeStampIndex = 0

            while (offset < frame.dataContent.size) {
                val x = TypeUtils.convertArrayToSignedInt(frame.dataContent, offset, step)
                offset += step
                val y = TypeUtils.convertArrayToSignedInt(frame.dataContent, offset, step)
                offset += step
                val z = TypeUtils.convertArrayToSignedInt(frame.dataContent, offset, step)
                offset += step
                accData.accSamples.add(AccSample(timeStamp = timeStamps[timeStampIndex], x = x, y = y, z = z))
                timeStampIndex++
            }
            return accData
        }

        private fun dataFromCompressedType0(frame: PmdDataFrame): AccData {
            //Note, special Wolfi type. See SAGRFC85.3
            val accData = AccData()
            val accFactor = frame.factor * 1000 // type 0 data arrives in G units, convert to milliG
            val samples = parseDeltaFramesAll(frame.dataContent, 3, 16, BlePMDClient.PmdDataFieldEncoding.SIGNED_INT)
            val timeStamps = PmdTimeStampUtils.getTimeStamps(previousFrameTimeStamp = frame.previousTimeStamp, frameTimeStamp = frame.timeStamp, samplesSize = samples.size, frame.sampleRate)
            for ((index, sample) in samples.withIndex()) {
                val x = (sample[0] * accFactor).toInt()
                val y = (sample[1] * accFactor).toInt()
                val z = (sample[2] * accFactor).toInt()
                accData.accSamples.add(AccSample(timeStamp = timeStamps[index], x = x, y = y, z = z))
            }
            return accData
        }

        private fun dataFromCompressedType1(frame: PmdDataFrame): AccData {
            val accData = AccData()
            val samples = parseDeltaFramesAll(frame.dataContent, TYPE_1_CHANNELS_IN_SAMPLE, TYPE_1_SAMPLE_SIZE_IN_BITS, BlePMDClient.PmdDataFieldEncoding.SIGNED_INT)
            val timeStamps = PmdTimeStampUtils.getTimeStamps(previousFrameTimeStamp = frame.previousTimeStamp, frameTimeStamp = frame.timeStamp, samplesSize = samples.size, frame.sampleRate)
            for ((index, sample) in samples.withIndex()) {
                val x = if (frame.factor != 1.0f) (sample[0].toFloat() * frame.factor).toInt() else sample[0]
                val y = if (frame.factor != 1.0f) (sample[1].toFloat() * frame.factor).toInt() else sample[1]
                val z = if (frame.factor != 1.0f) (sample[2].toFloat() * frame.factor).toInt() else sample[2]
                accData.accSamples.add(AccSample(timeStamp = timeStamps[index], x = x, y = y, z = z))
            }
            return accData
        }
    }
}

File: sources/iOS/ios-communications/Sources/iOSCommunications/ble/api/model/gatt/client/pmd/model/AccData.swift
//  Copyright © 2022 Polar. All rights reserved.

import Foundation

public class AccData {
    var timeStamp: UInt64 = 0
    
    public struct AccSample {
        let timeStamp: UInt64
        let x: Int32
        let y: Int32
        let z: Int32
    }
    
    var samples:[AccSample]
    
    init(timeStamp: UInt64 = 0, samples: [AccSample] = []) {
        self.timeStamp = timeStamp
        self.samples = samples
    }
    
    func parseFromFrame(frame: PmdDataFrame) throws {
        let data = try AccData.parseDataFromDataFrame(frame: frame)
        samples.append(contentsOf: data.samples)
    }
    
    private static let TYPE_0_SAMPLE_SIZE_IN_BYTES: UInt8 = 1
    private static let TYPE_0_SAMPLE_SIZE_IN_BITS: UInt8 = TYPE_0_SAMPLE_SIZE_IN_BYTES * 8
    private static let TYPE_0_CHANNELS_IN_SAMPLE: UInt8 = 3
    
    private static let TYPE_1_SAMPLE_SIZE_IN_BYTES: UInt8 = 2
    private static let TYPE_1_SAMPLE_SIZE_IN_BITS: UInt8 = TYPE_1_SAMPLE_SIZE_IN_BYTES * 8
    private static let TYPE_1_CHANNELS_IN_SAMPLE: UInt8 = 3
    
    private static let TYPE_2_SAMPLE_SIZE_IN_BYTES: UInt8 = 3
    private static let TYPE_2_SAMPLE_SIZE_IN_BITS: UInt8 = TYPE_2_SAMPLE_SIZE_IN_BYTES * 8
    private static let TYPE_2_CHANNELS_IN_SAMPLE: UInt8 = 3
    
    static func parseDataFromDataFrame(frame: PmdDataFrame) throws -> AccData {
        if (frame.isCompressedFrame) {
            switch (frame.frameType) {
            case PmdDataFrameType.type_0: return try dataFromCompressedType0(frame: frame)
            case PmdDataFrameType.type_1: return try dataFromCompressedType1(frame: frame)
            default: throw BleGattException.gattDataError(description: "Compressed FrameType: \(frame.frameType) is not supported by ACC data parser")
            }
        } else {
            switch (frame.frameType) {
            case PmdDataFrameType.type_0: return try dataFromRawType0(frame: frame)
            case PmdDataFrameType.type_1: return try dataFromRawType1(frame: frame)
            default: throw BleGattException.gattDataError(description: "Raw FrameType: \(frame.frameType) is not supported by ACC data parser")
            }
        }
    }
    
    private static func dataFromRawType0(frame: PmdDataFrame) throws -> AccData {
        var offset = 0
        let step = TYPE_0_SAMPLE_SIZE_IN_BYTES
        let samplesSize = Int(Double(frame.dataContent.count) / Double(step * TYPE_0_CHANNELS_IN_SAMPLE))
        let timeStamps = try PmdTimeStampUtils.getTimeStamps(previousFrameTimeStamp: frame.previousTimeStamp, frameTimeStamp: frame.timeStamp, samplesSize: UInt(samplesSize), sampleRate: frame.sampleRate)
        
        var timeStampIndex = 0
        var accSamples = [AccSample]()
        while offset < frame.dataContent.count {
            let x =  TypeUtils.convertArrayToSignedInt(frame.dataContent, offset: offset, size: Int(step))
            offset += Int(step)
            let y =  TypeUtils.convertArrayToSignedInt(frame.dataContent, offset: offset, size: Int(step))
            offset += Int(step)
            let z =  TypeUtils.convertArrayToSignedInt(frame.dataContent, offset: offset, size: Int(step))
            offset += Int(step)
            accSamples.append( AccSample( timeStamp: timeStamps[timeStampIndex], x: x, y: y, z: z))
            timeStampIndex += 1
        }
        return AccData(timeStamp: frame.timeStamp, samples: accSamples)
    }
    
    private static func dataFromRawType1(frame: PmdDataFrame) throws -> AccData {
        var offset = 0
        let step = TYPE_1_SAMPLE_SIZE_IN_BYTES
        let samplesSize = Int(Double(frame.dataContent.count) / Double(step * TYPE_1_CHANNELS_IN_SAMPLE))
        let timeStamps = try PmdTimeStampUtils.getTimeStamps(previousFrameTimeStamp: frame.previousTimeStamp, frameTimeStamp: frame.timeStamp, samplesSize: UInt(samplesSize), sampleRate: frame.sampleRate)
        
        var timeStampIndex = 0
        var accSamples = [AccSample]()
        while offset < frame.dataContent.count {
            let x =  TypeUtils.convertArrayToSignedInt(frame.dataContent, offset: offset, size: Int(step))
            offset += Int(step)
            let y =  TypeUtils.convertArrayToSignedInt(frame.dataContent, offset: offset, size: Int(step))
            offset += Int(step)
            let z =  TypeUtils.convertArrayToSignedInt(frame.dataContent, offset: offset, size: Int(step))
            offset += Int(step)
            accSamples.append( AccSample( timeStamp: timeStamps[timeStampIndex], x: x, y: y, z: z))
            timeStampIndex += 1
        }
        return AccData(timeStamp: frame.timeStamp, samples: accSamples)
    }
    
    private static func dataFromCompressedType0(frame: PmdDataFrame) throws -> AccData {
        //Note, special Wolfi type. See SAGRFC85.3
        let samples = Pmd.parseDeltaFramesToSamples(frame.dataContent, channels: 3, resolution: 16)
        let timeStamps = try PmdTimeStampUtils.getTimeStamps(previousFrameTimeStamp: frame.previousTimeStamp, frameTimeStamp: frame.timeStamp, samplesSize: UInt(samples.count), sampleRate: frame.sampleRate)
        let accFactor = frame.factor * 1000 // type 0 data arrives in G units, convert to milliG
        
        var accSamples = [AccSample]()
        for (index, sample) in samples.enumerated() {
            accSamples.append( AccSample( timeStamp: timeStamps[index],
                                          x: Int32((Float(sample[0]) * accFactor)),
                                          y: Int32((Float(sample[1]) * accFactor)),
                                          z: Int32((Float(sample[2]) * accFactor))
                                        ))
        }
        return AccData(timeStamp: frame.timeStamp, samples: accSamples)
    }
    
    private static func dataFromCompressedType1(frame: PmdDataFrame) throws -> AccData {
        let samples = Pmd.parseDeltaFramesToSamples(frame.dataContent, channels: TYPE_1_CHANNELS_IN_SAMPLE, resolution: TYPE_1_SAMPLE_SIZE_IN_BITS)
        let timeStamps = try PmdTimeStampUtils.getTimeStamps(previousFrameTimeStamp: frame.previousTimeStamp, frameTimeStamp: frame.timeStamp, samplesSize: UInt(samples.count), sampleRate: frame.sampleRate)
        
        var accSamples = [AccSample]()
        for (index, sample) in samples.enumerated() {
            let x:Int32
            let y:Int32
            let z:Int32
            if (frame.factor != 1.0){
                x = Int32((Float(sample[0]) * frame.factor))
                y = Int32((Float(sample[1]) * frame.factor))
                z = Int32((Float(sample[2]) * frame.factor))
            } else {
                x = sample[0]
                y = sample[1]
                z = sample[2]
            }
            accSamples.append( AccSample( timeStamp: timeStamps[index], x: x, y: y, z: z))
        }
        return AccData(timeStamp: frame.timeStamp, samples: accSamples)
    }
}
